pricegraph/src/graph/shortest_paths.rs

```diff
@@ -0,0 +1,311 @@
+        // This is the last step of the Bellman-Ford algorithm. It tries to relax
+        // each node: if a node can be relaxed then a negative cycle exists and is
+        // created in the predecessor store; otherwise no such cycle exists.
+
+        for i in self.graph.node_identifiers() {
+            for edge in self.graph.edges(i) {
+                let j = edge.target();
+                let w = *edge.weight();
+
+                if *self.distance(i) + w < *self.distance(j) {
+                    self.update_predecessor(j, Some(i));
+                    return Some(j);
+                }
+            }
+        }
+        None
+    }
+
+    /// Returns a negative cycle, if it exists.
+    fn find_cycle(&mut self) -> Option<NegativeCycle<G::NodeId>> {
+        let search_start = match self.mark_cycle() {
+            Some(node) => node,
+            None => return None,
+        };
+
+        find_cycle(
+            self.graph,
+            &self.predecessor_store.predecessors,
+            search_start,
+            None,
+        )
+    }
+}
+
+impl<G> ShortestPathGraph<G>
+where
+    G: NodeCount + IntoNodeIdentifiers + IntoEdges + NodeIndexable,
+    G::EdgeWeight: FloatMeasure,
+{
+    /// Creates a representation of all shortest paths from the given source
+    /// to any other node in the graph.
+    ///
+    /// Shortest paths are well defined if and only if the graph does not
+    /// contain any negative weight cycle reachabe from the source. If a
+    /// negative weight cycle is detected, it is returned as an error.
+    pub fn new(g: G, source: G::NodeId) -> Result<Self, NegativeCycle<G::NodeId>> {
+        bellman_ford(g, source)
+    }
+}
+
+/// This implementation follows closely the one that can be found in the
+/// `petgraph` crate, but it has customized output types.
+///
+/// The orginal source can be found here:
+/// https://docs.rs/petgraph/0.5.0/src/petgraph/algo/mod.rs.html#745-792
+fn bellman_ford<G>(
+    g: G,
+    source: G::NodeId,
+) -> Result<ShortestPathGraph<G>, NegativeCycle<G::NodeId>>
+where
+    G: NodeCount + IntoNodeIdentifiers + IntoEdges + NodeIndexable,
+    G::EdgeWeight: FloatMeasure,
+{
+    let mut shortest_path_graph = ShortestPathGraph::empty(g, source);
+
+    // scan up to |V| - 1 times.
+    for _ in 1..g.node_count() {
+        let mut did_update = false;
+        for i in g.node_identifiers() {
+            for edge in g.edges(i) {
+                let i = edge.source();
+                let j = edge.target();
+                let w = *edge.weight();
+                if *shortest_path_graph.distance(i) + w < *shortest_path_graph.distance(j) {
+                    shortest_path_graph.update_distance(j, *shortest_path_graph.distance(i) + w);
+                    shortest_path_graph.update_predecessor(j, Some(i));
+                    did_update = true;
+                }
+            }
+        }
+        if !did_update {
+            break;
+        }
```

Comment:
I find this implementation nice to read. Maybe we should measure the performance impact of this.